### PR TITLE
Remove function use in early steps

### PR DESCRIPTION
**Pull request by drwells**
In the process of (hopefully) closing #650 I also cleaned up these tutorials a bit.

My primary concern is that we lose a nice description of what happens when an exception is thrown from `Function::value_list` because the two arguments have different lengths. I took it out because that function call no longer exists. Thoughts?

### CODE REVIEW

File: examples/step-6/doc/results.dox [hunk 1]
```diff
@@ -14,19 +14,19 @@ Cycle 2:
...
 Cycle 6:
    Number of active cells:       1916
-   Number of degrees of freedom: 9113
+   Number of degrees of freedom: 8993
 Cycle 7:
-   Number of active cells:       3884
-   Number of degrees of freedom: 18401
+   Number of active cells:       3860
+   Number of degrees of freedom: 18353
 @endcode
```
Comment (tamiko) [inline examples/step-6/doc/results.dox]: When did the number of dofs change? :-D

File: examples/step-7/step-7.cc [hunk 2]
```diff
@@ -64,11 +64,7 @@
 // the same file as the FEValues class:
 #include <deal.II/fe/fe_values.h>
 
-// We need one more include from standard C++, which is necessary when we try
-// to find out the actual type behind a pointer to a base class. We will
-// explain this in slightly more detail below. The other two include files are
-// obvious then:
-#include <typeinfo>
+// This is C++:
```
Comment (tamiko) [inline examples/step-7/step-7.cc]: What about just removing the comment entirely? I tried to understand what you meant by "this is C++" now for 3 minutes.

File: examples/step-7/step-7.cc [hunk 3]
```diff
@@ -171,17 +167,25 @@ namespace Step7
...
+  // library provides a suitable class for this. One new thing about this
+  // class is that it explicitly uses the Tensor objects, which previously
+  // appeared as intermediate terms in step-3 and step-4. A tensor is a
+  // generalization of scalars (rank zero tensors), vectors (rank one
+  // tensors), and matrices (rank two tensors), as well as higher dimensional
+  // objects. The Tensor class requires two template arguments: the tensor
+  // rank and tensor dimension. For example, here we use tensors of rank one
+  // (vectors) with dimension <code>dim</code> (so they have <code>dim</code>
+  // entries.) While this is a bit less flexible than using Vector, the
+  // compiler can generate much, much faster code when the length of the
```
Comment (tamiko) [inline examples/step-7/step-7.cc]: What about removing "much, much", just "... can generate faster code ...".

**Comment by tamiko** [comment]
I like the cleanup.

File: examples/step-7/step-7.cc (hunk 2 above)
Comment (drwells) [inline examples/step-7/step-7.cc]: Some of the other tutorial programs (5, 9, and 17, among others) add that text before including C++ system headers, so I added it here too. It doesn't really say anything so I can remove it.

**Comment by drwells** [comment]
I force pushed fixes to the two things pointed out above.

File: examples/step-6/doc/results.dox (hunk 1 above)
Comment (drwells) [inline examples/step-6/doc/results.dox]: No idea. It looks like we added the result files to the repository in 2006 so, as far as I can tell, `results.dox` could have data from when it was added in 2000.

**Comment by tamiko** [comment]
Thanks! I'll give others some time to comment.

**Comment by tamiko** [comment]
@drwells Please rebase and merge.

**Comment by drwells** [comment]
I am traveling for the next few days; I will handle this on Monday.

On Friday, July 22, 2016, Matthias Maier <notifications@github.com> wrote:
> @drwells Please rebase and merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci5.googleusercontent.com/proxy/T3a-Ckx10l9GvpMOmocOzw4nv1HGzG9GxhoOMvxWIRoXbPAvy302FwH2yfBNyNmlvZkWeONw2Egpm3CTEkI3nNzBBXBXewz3yctSmULwzZ5sch4F7FiE7_Fgf5OMz7RaGy742ZttXphTgLOJzvWXbf1saZ5utQ=s0-d-e1-ft#https://github.com/notifications/beacon/ADvE6-3Kq-rf0Dp3vgC3XUXDROz_W-0yks5qYKeHgaJpZM4JOUWl.gif>

File: examples/step-8/step-8.cc [hunk 4]
```diff
@@ -208,77 +155,24 @@ namespace Step8
...
-    // with it. At the same time, using virtual function call bypassing, the
-    // code is no less efficient than if we had written it twice in the first
-    // place:
-    for (unsigned int p=0; p<n_points; ++p)
-      RightHandSide<dim>::vector_value (points[p],
-                                        value_list[p]);
+        // Likewise, if <code>points[point_n]</code> is in the vicinity of the
+        // origin, then set the y-force to 1, otherwise to zero:
+        if (points[point_n].norm_square() < 0.2*0.2)
+          values[point_n][1] = 1;
```
Comment (tjhei) [inline examples/step-8/step-8.cc]: I would prefer "1." and "0." here.

**Comment by tjhei** [comment]
excellent work! :+1:

**Comment by drwells** [comment]
@tjhei I made the suggested change (and I did the same thing in the block above) and also added a changelog entry.

File: examples/step-5/step-5.cc [hunk 5]
```diff
@@ -569,7 +402,43 @@ void Step5<dim>::run ()
...
   // arbitrary dimension. Since this is only a demonstration program, we will
   // not use different input files for the different dimensions, but rather
-  // kill the whole program if we are not in 2D:
+  // kill the whole program if we are not in 2D. Of course, since the main
+  // function below assumes that we are working in two dimensions we could
+  // skip this check, in this version of the program, without any ill effects.
+  //
+  // It turns out that more than 90 per cent of programming errors are invalid
+  // function parameters such as invalid array sizes, etc, so we use
+  // assertions heavily through out deal.II to catch such mistakes. For this,
```
Comment (bangerth) [inline examples/step-5/step-5.cc]: -> throughout

File: examples/step-5/step-5.cc [hunk 6]
```diff
@@ -569,7 +402,43 @@ void Step5<dim>::run ()
...
+  // the <code>Assert</code> macro is a good choice, since it makes sure that
+  // the condition which is given as first argument is valid, and if not
+  // throws an exception (its second argument) which will usually terminate
+  // the program giving information where the error occurred and what the
+  // reason was. This generally reduces the time to find programming errors
+  // dramatically and we have found assertions an invaluable means to program
+  // fast.
```
Comment (bangerth) [inline examples/step-5/step-5.cc]: quickly?

File: examples/step-5/step-5.cc [hunk 7]
```diff
@@ -569,7 +402,43 @@ void Step5<dim>::run ()
...
+  //
+  // On the other hand, all these checks (there are over 9000 of them in the
+  // library at present) should not slow down the program too much if you want
+  // to do large computations. To this end, the <code>Assert</code> macro is
+  // only used in debug mode and expands to nothing if in optimized
+  // mode. Therefore, while you test your program on small problems and debug
+  // it, the assertions will tell you where the problems are. Once your
+  // program is stable, you can switch off debugging and the program will run
+  // your real computations without the assertions and at maximum speed. (In
+  // fact, it turns out the switching off all the checks in the library that
```
Comment (bangerth) [inline examples/step-5/step-5.cc]: that switching off?

The sentence is long and convoluted. Feel free to rephrase.

File: examples/step-7/step-7.cc [hunk 8]
```diff
@@ -171,17 +166,25 @@ namespace Step7
...
+  // at compile time. Additionally, specifying a Tensor of rank one and
+  // dimension <code>dim</code> guarantees that the tensor will have the right
+  // shape (since it is built into the type of the object itself), so the
+  // compiler can catch most size-related mistakes for us.
   template <int dim>
   class Solution : public Function<dim>,
     protected SolutionBase<dim>
   {
   public:
-    Solution () : Function<dim>() {}
```
Comment (bangerth) [inline examples/step-7/step-7.cc]: This was necessary at one point for some version of the intel compiler. Let's suppose they fixed their compiler in the meantime, then the constructor can indeed go.

**Comment by bangerth** [comment]
I like this better than the old approach of gratuitously packaging everything into `Function` objects. That said, there are places where one needs to do it (e.g., when interpolating initial or boundary conditions). Where is the first place where we now do this, and do we give a brief introduction to the `Function` class there?

**Comment by tamiko** [comment]
@drwells @bangerth 
> Where is the first place where we now do this, and do we give a brief introduction to the Function class there?

I suggest we improve this in a follow-up PR.

**Comment by drwells** [comment]
Sorry for the delay. I tried to get rid of the extra default constructors and some weird things happened:

1. By the standard (see http://en.cppreference.com/w/cpp/language/default_constructor) `FunctionTime` and `Function` are default constructible
2. everything is fine with GCC6 in C++11 and C++14 modes
3. Any step where I deleted the default constructor (e.g., `step-13`) for things inheriting from `Function` now fails to compile with clang-3.8 in C++14 mode, claiming that no default constructor is available (more specifically: that one is trying to create a `const` instance of something with unitialized fields). This has to be a bug since `std::is_default_constructible<const RightHandSide<dim>>::value` is `1` even when clang refuses to compile things (one can check this easily by writing a 50-line program with something that inherits from `Function` in it).

I need to go back and make sure that all of the 'unneeded' constructors are present (enough that clang is happy). I'll try to get to this tonight.

**Comment by bangerth** [comment]
Ah, what a bummer. Might be worth opening a bug report with the clang people.

**Comment by drwells** [comment]
@bangerth We currently have a nice introduction of the `Function` class in `step-4`. I will add a bit of explanation on the class there.

**Comment by bangerth** [comment]
Ah, alright! Let us know when you think you're through and want this merged!

**Comment by drwells** [comment]
That should do it.

I did include an introduction of `Function` in step-4; it only took about a minute to add since I knew where to put it.

**Comment by bangerth** [comment]
Great, thanks for the hard work on this!

**Pull request merged**